### PR TITLE
options to build sk p3 and shoc separately

### LINKED ISSUE #2963: split small kernels into two distinct cmake logic
p3 sk -----> 2x slowdown
shoc sk ---> 2x speedup

### PR DESCRIPTION
**Pull request by mahf708**
I opted to keep SCREAM_SMALL_KERNELS for _extreme_ backward compatibility, but I am happy to remove it altogether. Please advise during review.

Fixes #2963

### CODE REVIEW

**Review by jgfouca — DISMISSED** [review]

File: cime_config/tests.py [hunk 1]
```diff
@@ -650,6 +650,7 @@
             "SMS_D_Ln9.ne4_ne4.F2010-SCREAMv1-noAero.scream-output-preset-3",
             "ERP_Ln22.ne4pg2_ne4pg2.F2010-SCREAMv1.scream-output-preset-4",
             "ERS_D_Ln22.ne4pg2_ne4pg2.F2010-SCREAMv1.scream-rad_frequency_2--scream-output-preset-5",
+            "ERS_Ln22.ne4pg2_ne4pg2.F2010-SCREAMv1.scream-small_kernels--scream-output-preset-5",
```
Comment (bartgol) [inline cime_config/tests.py:653]: Did you mean to use `small_kernels_p3` (or shoc) maybe? To test the single sk choice?

File: components/eamxx/src/physics/p3/impl/p3_main_impl.hpp [hunk 2]
```diff
@@ -349,7 +349,7 @@ ::p3_main(
   Int nk,
   const physics::P3_Constants<S> & p3constants)
 {
-#ifndef SCREAM_SMALL_KERNELS
+#ifndef SCREAM_P3_SMALL_KERNELS
```
Comment (bartgol) [inline components/eamxx/src/physics/p3/impl/p3_main_impl.hpp:352]: Maybe for another PR, but: how do ppl feel about swapping the code to have _positive_ CPP checks? That is
```c++
#ifdef SCREAM_P3_SMALL_KERNELS
  // do small kernels
#else
  // do monolithic
#endif
```
? My brain always has to use a few extra cycles to parse the negative, especially considering that during the first parse it never registers the `n` in the CPP pragma...

File: components/eamxx/src/scream_config.h.in [hunk 3]
```diff
@@ -45,8 +45,10 @@
 // Whether or not to run RRTMGP debug checks
 #cmakedefine SCREAM_RRTMGP_DEBUG
 
-// Whether monolithic kernels are on
+// Whether or not small kernels are on
```
Comment (bartgol) [inline components/eamxx/src/scream_config.h.in:48]: We should prob be more specific. This could be OFF, and yet p3 may use small kernels. So maybe I would rephrase as "Whether or not small kernels are used in ALL targets that support them".

**Review by bartgol — COMMENTED** [review]
Looks good. I have one comment on a shell mod.

Re: keep/drop SCREAM_SMALL_KERNELS. Personally, I would keep it, but change its name to something that hints at the fact that it enables SK _everywhere_ we can. Something like SCREAM_PREFER_SMALL_KERNELS or SCREAM_ENABLE_ALL_SMALL_KERNELS. Meanwhile, we can keep it as is.

File: cime_config/tests.py (hunk 1 above)
Comment (bartgol) [reply cime_config/tests.py:653]: It's ok to have 3 testmods. I was hoping you would test one of those two in this PR, so that we can check (in logs) that indeed we use it in, say, shoc but not p3...

File: cime_config/tests.py [hunk 4]
```diff
@@ -650,6 +650,9 @@
...
+            "ERS_Ln22.ne4pg2_ne4pg2.F2010-SCREAMv1.scream-small_kernels_p3--scream-output-preset-5",
```
Comment (bartgol) [inline cime_config/tests.py:654]: Maybe not for now, but I wonder if we could add these to other existing tests, and keep the v1 lowres suite compact. I like that it's currently a handful tests (runs quickly). Yes, they're all ne4 tests, so they run fast, but I'm also thinking about the build time...

@jgfouca Can we make all these tests use the same sharedlib phase? I know we can make tests use the same exec, but that may not be possible here,  since the small kernels are build time differences (although the first 4 tests could maybe be made to use the same exec?).

**Review by bartgol — DISMISSED** [review]

File: components/eamxx/CMakeLists.txt [hunk 5]
```diff
@@ -198,8 +198,10 @@ set(NetCDF_C_PATH ${DEFAULT_NetCDF_C_PATH} CACHE FILEPATH "Path to netcdf C inst
 set(SCREAM_MACHINE ${DEFAULT_SCREAM_MACHINE} CACHE STRING "The CIME/SCREAM name for the current machine")
 option(SCREAM_MPI_ON_DEVICE "Whether to use device pointers for MPI calls" ON)
 option(SCREAM_ENABLE_MAM "Whether to enable MAM aerosol support" ON)
-set(SCREAM_SMALL_KERNELS ${DEFAULT_SMALL_KERNELS} CACHE STRING "Use small, non-monolothic kokkos kernels")
-if (NOT SCREAM_SMALL_KERNELS)
+set(SCREAM_SMALL_KERNELS ${DEFAULT_SMALL_KERNELS} CACHE STRING "Use small, non-monolothic kokkos kernels for all components")
```
Comment (tcclevenger) [inline components/eamxx/CMakeLists.txt:201]: Like you did with `#cmakedefine` below, maybe say "...all components that have support" or something like that since not all components have a small kernel distinction

File: components/eamxx/src/physics/p3/impl/p3_main_impl.hpp (hunk 2 above)
Comment (tcclevenger) [reply components/eamxx/src/physics/p3/impl/p3_main_impl.hpp:352]: I also prefer positive!

**Review by tcclevenger — COMMENTED** [review]
Just a minor comment.

File: components/eamxx/CMakeLists.txt (hunk 5 above)
Comment (tcclevenger) [reply components/eamxx/CMakeLists.txt:201]: Works for me.

File: components/eamxx/src/physics/p3/impl/p3_main_impl.hpp (hunk 2 above)
Comment (ndkeen) [reply components/eamxx/src/physics/p3/impl/p3_main_impl.hpp:352]: can we make a rule that only positives are allowed from here on?

**Review by ndkeen — APPROVED** [review]
approving without testing -- i'm sure we can make it work

File: components/eamxx/src/physics/p3/impl/p3_main_impl.hpp (hunk 2 above)
Comment (bartgol) [reply components/eamxx/src/physics/p3/impl/p3_main_impl.hpp:352]: I do not disagree.

**Review by tcclevenger — APPROVED** [review]

**Comment by jgfouca** [comment]
@mahf708 , yeah, just ping me if the RETEST looks OK.

**Comment by tcclevenger** [comment]
Do we expect the SK CIME cases to diff?
